### PR TITLE
Metrics wrapper part 2

### PR DESCRIPTION
**Pull request by zcain117**
Finish the main functionality of metrics_test_wrapper.py by retrieving past metrics and performing regression checks for the latest run of a subprocess.

Tested by running locally

### CODE REVIEW

File: torch_xla/test/metrics_test_wrapper.py [hunk 1]
```diff
@@ -112,21 +150,69 @@ def _write_to_disk(output_string, output_filename):
...
+      ordered_config_dicts.insert(0, json.load(f))
+    except google.api_core.exceptions.NotFound:
+      pass
+    if path_to_search == '':
+      break
+    path_to_search = os.path.split(path_to_search)[0]
+  if not ordered_config_dicts:
+    print('No config files found. See example usage at top of '
+          'metrics_test_wrapper.py')
+    sys.exit(1)  # Return non-OK status since config is required.
```
Comment (dlibenzi) [inline torch_xla/test/metrics_test_wrapper.py]: Module functions should not exit().

Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: What is the preferred way to indicate a non-0 exit code so that the caller of this script knows something is wrong?

Comment (dlibenzi) [reply torch_xla/test/metrics_test_wrapper.py]: Module functions should `raise` on errors.

Comment (dlibenzi) [reply torch_xla/test/metrics_test_wrapper.py]: You `raise` with proper error message, not printing it.
If you want to have different exit codes depending on exception types, you have an try:catch: in main().

Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: I will switch to `raise`, but for my understanding: is this considered a module function if it occurs inside `if __name__ == '__main__':` ?

**Comment by dlibenzi** [comment]
Oh, sorry, wait. This is actually a main().
I though it was a module, whose functions were called from a main script.
Why does the main live in pytorch_xla/test/ ?
Usually we have main()s in scripts, and modules providing support functions in pytorch_xla/...

**Comment by zcain117** [comment]
> Oh, sorry, wait. This is actually a main().
> I though it was a module, whose functions were called from a main script.
> Why does the main live in pytorch_xla/test/ ?
> Usually we have main()s in scripts, and modules providing support functions in pytorch_xla/...

I can move this to `scripts/` or to `test/`. Another wrapper script we use with a __main__ lives in torch_xla/distributed: https://github.com/pytorch/xla/blob/master/torch_xla/distributed/xla_dist.py#L728

**Comment by dlibenzi** [comment]
OK, leave it here.

File: torch_xla/test/metrics_test_wrapper.py [hunk 2]
```diff
@@ -112,21 +150,69 @@ def _write_to_disk(output_string, output_filename):
...
+    try:
+      f = gcsfs.open(os.path.join(FLAGS.root, path_to_search, 'config.json'))
+      ordered_config_dicts.insert(0, json.load(f))
+    except google.api_core.exceptions.NotFound:
+      pass
+    if path_to_search == '':
+      break
+    path_to_search = os.path.split(path_to_search)[0]
+  if not ordered_config_dicts:
+    print('No config files found. See example usage at top of '
```
Comment (dlibenzi) [inline torch_xla/test/metrics_test_wrapper.py]: These error-prints should use file=sys.stderr.

**Comment by zcain117** [comment]
Switched to use `raise` instead of `sys.exit` and use stderr for the 1 remaining error-print

File: torch_xla/test/metrics_test_wrapper.py [hunk 3]
```diff
@@ -112,21 +152,72 @@ def _write_to_disk(output_string, output_filename):
   if not FLAGS.root or not FLAGS.test_folder_name:
```
Comment (dlibenzi) [inline torch_xla/test/metrics_test_wrapper.py]: If these are required, I'd use required arguments.
Example:

https://github.com/pytorch/xla/blob/4a33f7ce62f3da58690e83c4e3ad97871d00e1f7/scripts/gen.py#L1036

File: torch_xla/test/metrics_test_wrapper.py [hunk 4]
```diff
@@ -112,21 +152,72 @@ def _write_to_disk(output_string, output_filename):
...
+  ordered_config_dicts = []
+
+  path_to_search = FLAGS.test_folder_name
+  while True:
+    try:
+      f = gcsfs.open(os.path.join(FLAGS.root, path_to_search, 'config.json'))
+      ordered_config_dicts.insert(0, json.load(f))
+    except google.api_core.exceptions.NotFound:
+      pass
+    if path_to_search == '':
```
Comment (dlibenzi) [inline torch_xla/test/metrics_test_wrapper.py]: `if not path_to_search:`

File: torch_xla/test/metrics_test_wrapper.py [hunk 5]
```diff
@@ -112,21 +152,72 @@ def _write_to_disk(output_string, output_filename):
...
+    output_string = '{}\n\n{}'.format(FLAGS, metrics)
+    output_filename = os.path.join(
+        metrics_storage_dir, datetime.datetime.utcnow().strftime('%Y_%m_%d'))
+    _write_to_disk(output_string, output_filename)
+
+  if regression_report:
+    raise AssertionError(
+        'Non-empty XLA metrics regression report:\n{}'.format(
+            regression_report))
+  if sp_return_code:
```
Comment (dlibenzi) [inline torch_xla/test/metrics_test_wrapper.py]: Shouldn't we bail out immediately upon process failure?

File: torch_xla/test/metrics_test_wrapper.py (hunk 3 above)
Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: Added `required=True` to both args and removed the ValueError

File: torch_xla/test/metrics_test_wrapper.py (hunk 4 above)
Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: Done

File: torch_xla/test/metrics_test_wrapper.py (hunk 5 above)
Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: I lean toward no but could go either way. Any immediate subprocess failure will result in no XLA metrics file, which leads to a fail fast Exception in `_find_correct_metrics_file`.

I think the most common case of non-zero subprocess return code will be failing the accuracy assertion, in which case we still want to save metrics and check for metrics regressions.

The bad case is when we fail midway through training since then we will save metrics for partial runs - but hopefully we will see the alert when the training job died during a partial run and we could clean up the partial metrics? 

Or maybe I could find a way to pass the number of trained epochs from the child process to this wrapper and only write/assert XLA metrics if we trained for all the desired epochs

Comment (dlibenzi) [reply torch_xla/test/metrics_test_wrapper.py]: But if the process exits early with error, there is little point in comparing the metrics I think.
They might be totally off due to truncation.

Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: Yes exactly, so 3 cases:
1. process exits non-0 before training starts - **no current problem** since there's no metrics file
2. process exits non-0 after training finishes since accuracy too low - **no current problem** since desired behavior is to check XLA metrics and write them to disk
3. process exits non-0 during training - **this is a problem** since the metrics are not to be trusted.

Some solutions:
-  use one exit code for case 2 and a different one for case 3 in the child process
- parse stdout logs of child process to see if training finished or not or maybe to see if number of completed epochs == desired number of epochs
- add another field to XLA metrics like "finished training"
- add number of epochs to child process tensorboard summary and search for the summary in this wrapper to see if we completed all epochs

Comment (zcain117) [reply torch_xla/test/metrics_test_wrapper.py]: Oh and another solution option:
Don't check or write metrics at all unless subprocess has exit code 0. This just means we lose metrics if accuracy regresses - which is hopefully rare

**Comment by zcain117** [comment]
Switched to raise early for non-0 child process exit code. Probably better to occasionally miss useful metrics data (if child fails only due to accuracy) than to have save bad metrics data if child dies partway through training

**Comment by dlibenzi** [comment]
I'd try to make that big main code, into functions.
For reference, look at this:

https://github.com/pytorch/xla/blob/master/scripts/debug_run.py

Look how work is split into well defined functions, and single functions (or body of code) are not much long.
Functions will also make that code more of a module, increasing re-usability.
